invokeai/backend/model_manager/config.py

```diff
@@ -0,0 +1,342 @@
+        """Update the object with fields in dict."""
+        for key, value in attributes.items():
+            setattr(self, key, value)  # may raise a validation error
+
+
+class CheckpointConfig(ModelConfigBase):
+    """Model config for checkpoint-style models."""
+
+    model_format: Literal[ModelFormat.Checkpoint] = ModelFormat.Checkpoint
+    config: str = Field(description="path to the checkpoint model config file")
+
+
+class DiffusersConfig(ModelConfigBase):
+    """Model config for diffusers-style models."""
+
+    model_format: Literal[ModelFormat.Diffusers] = ModelFormat.Diffusers
+
+
+class LoRAConfig(ModelConfigBase):
+    """Model config for LoRA/Lycoris models."""
+
+    model_format: Literal[ModelFormat.Lycoris, ModelFormat.Diffusers]
+
+
+class VaeCheckpointConfig(ModelConfigBase):
+    """Model config for standalone VAE models."""
+
+    model_format: Literal[ModelFormat.Checkpoint] = ModelFormat.Checkpoint
+
+
+class VaeDiffusersConfig(ModelConfigBase):
+    """Model config for standalone VAE models (diffusers version)."""
+
+    model_format: Literal[ModelFormat.Diffusers] = ModelFormat.Diffusers
+
+
+class ControlNetDiffusersConfig(DiffusersConfig):
+    """Model config for ControlNet models (diffusers version)."""
+
+    model_format: Literal[ModelFormat.Diffusers] = ModelFormat.Diffusers
+
+
+class ControlNetCheckpointConfig(CheckpointConfig):
+    """Model config for ControlNet models (diffusers version)."""
+
+    model_format: Literal[ModelFormat.Checkpoint] = ModelFormat.Checkpoint
+
+
+class TextualInversionConfig(ModelConfigBase):
+    """Model config for textual inversion embeddings."""
+
+    model_format: Literal[ModelFormat.EmbeddingFile, ModelFormat.EmbeddingFolder]
+
+
+class MainConfig(ModelConfigBase):
+    """Model config for main models."""
+
+    vae: Optional[str] = Field(None)
+    variant: ModelVariantType = ModelVariantType.Normal
+
+
+class MainCheckpointConfig(CheckpointConfig, MainConfig):
+    """Model config for main checkpoint models."""
+
+    config: str
+
+
+class MainDiffusersConfig(DiffusersConfig, MainConfig):
+    """Model config for main diffusers models."""
+
+
+class ONNXSD1Config(MainConfig):
+    """Model config for ONNX format models based on sd-1."""
+
+    model_format: Literal[ModelFormat.Onnx, ModelFormat.Olive]
+
+
+class ONNXSD2Config(MainConfig):
+    """Model config for ONNX format models based on sd-2."""
+
+    model_format: Literal[ModelFormat.Onnx, ModelFormat.Olive]
+    # No yaml config file for ONNX, so these are part of config
+    prediction_type: SchedulerPredictionType
+    upcast_attention: bool
+
+
+class IPAdapterConfig(ModelConfigBase):
+    """Model config for IP Adaptor format models."""
+
+    model_format: Literal[ModelFormat.InvokeAI]
+
+
+AnyModelConfig = Union[
+    MainCheckpointConfig,
+    MainDiffusersConfig,
+    LoRAConfig,
+    TextualInversionConfig,
+    ONNXSD1Config,
+    ONNXSD2Config,
+    ModelConfigBase,
+]
```

Comment:
Is this is missing VAE, ControlNet and IPAdapter? FWIW, the schemas and model varieties picked up by OpenAPI are not missing those 3.

Reply:
Fixed. The original pattern was for each of the methods that returned a config to return a `ModelConfigBase`, but mypy didn't like that. Passing these unions to pydantic's `parse_obj_as()` method to turn them into the right subclass still seems like a weird and ugly thing to do.

Reply:
Pydantic will test the object against each of the models and go with the first one that fits the type (ie "duck typing"). This can lead to undesirable behaviour.

The right way to do this in pydantic is using an annotated field with an explicit discriminator: https://docs.pydantic.dev/2.4/api/standard_library_types/#discriminated-unions-aka-tagged-unions

It doesn't do any duck-typing and instead instantiates the model with a matching discriminator. If none exists, you get a validatior error.

Note that nested unions work, and actually make a lot of sense for our model configs. Also note that those docs are for pydantic v2 so YMMV pending #4758 

Functional discriminators would make this much simpler for models without a single discriminator field, and is and active WIP for pydantic v2: https://github.com/pydantic/pydantic/issues/7462